src/libstore/gc.cc

```diff
@@ -774,44 +782,21 @@ void LocalStore::collectGarbage(const GCOptions & options, GCResults & results)
             printInfo("determining live/dead paths...");
 
         try {
-
             AutoCloseDir dir(opendir(realStoreDir.get().c_str()));
             if (!dir) throw SysError("opening directory '%1%'", realStoreDir);
 
-            /* Read the store and immediately delete all paths that
-               aren't valid.  When using --max-freed etc., deleting
-               invalid paths is preferred over deleting unreachable
-               paths, since unreachable paths could become reachable
-               again.  We don't use readDirectory() here so that GCing
-               can start faster. */
+            /* Read the store and delete all paths that are invalid or
+               unreachable. We don't use readDirectory() here so that
+               GCing can start faster. */
```

Comment:
What's the strategy for making sure clients can safely add a gc root for the paths they create during the gc process?

I expected the store directory to be read at the very start, to limit the deletable paths to those that existed at the start. That way only one bad interleaving remains: client/addToStore -> gc/readdir -> gc/findRoots -> client/addRoot, but this seems already far less likely than client/addToStore -> gc/readdir -> client/addRoot -> gc/findRoots.
Ideally we'd have implicit temporary roots per daemon connection or per LocalStore object.

> We don't use readDirectory() here so that GCing can start faster.

It's a synchronous interleaving, so any real speedup would have to come from kernel read-ahead cleverness. If you start it early in the gc process, you could run it concurrently.

Reply:
> What's the strategy for making sure clients can safely add a gc root for the paths they create during the gc process?

Clients connect to the garbage collector to inform it about temporary roots.